dynamiq/memory/backends/weaviate.py

```diff
@@ -0,0 +1,372 @@
+import time
+import uuid
+from typing import Any, ClassVar
+
+from pydantic import ConfigDict, Field, PrivateAttr
+
+from dynamiq.connections import Weaviate as WeaviateConnection
+from dynamiq.memory.backends.base import MemoryBackend
+from dynamiq.nodes.embedders.base import DocumentEmbedder, DocumentEmbedderInputSchema
+from dynamiq.prompts import Message, MessageRole
+from dynamiq.storages.vector.weaviate import WeaviateVectorStore, WeaviateWriterVectorStoreParams
+from dynamiq.types import Document
+from dynamiq.utils.logger import logger
+
+
+class WeaviateMemoryError(Exception):
+    """Base exception class for Weaviate Memory Backend errors."""
+
+    pass
+
+
+class Weaviate(MemoryBackend):
+    """
+    Weaviate implementation of the memory storage backend.
+
+    Uses WeaviateVectorStore to manage documents (messages) in a Weaviate collection.
+    Leverages vector embeddings for semantic search and metadata for filtering
+    and chronological ordering.
+    """
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    name: str = "Weaviate"
+    connection: WeaviateConnection = Field(default_factory=WeaviateConnection)
+    embedder: DocumentEmbedder
+    index_name: str = Field(default="conversations")
+    tenant_name: str | None = Field(default=None)
+    create_if_not_exist: bool = Field(default=True)
+    content_key: str = Field(default="message_content")
+    alpha: float = Field(default=0.5, description="Alpha for hybrid search (0=keyword, 1=vector)")
+
+    _vector_store: WeaviateVectorStore | None = PrivateAttr(default=None)
+
+    _ROLE_KEY: ClassVar[str] = "message_role"
+    _TIMESTAMP_KEY: ClassVar[str] = "message_timestamp"
+    _MESSAGE_ID_KEY: ClassVar[str] = "message_id"
+
+    _CORE_MEMORY_PROPERTIES: ClassVar[list[str]] = [
+        "message_role",
+        "message_timestamp",
+        "message_id",
+        "user_id",
+        "session_id",
+        "_original_id",
+    ]
+
+    @property
+    def to_dict_exclude_params(self) -> dict[str, bool]:
+        """Define parameters to exclude during serialization."""
+        return super().to_dict_exclude_params | {
+            "embedder": True,
+            "_vector_store": True,
```

Comment:
It seems private attributes excluded automatically during serialization. Please confirm